libsolidity/analysis/ImmutableValidator.cpp

```diff
@@ -0,0 +1,271 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libsolidity/analysis/ImmutableValidator.h>
+
+#include <libsolutil/CommonData.h>
+
+#include <boost/range/adaptor/reversed.hpp>
+
+using namespace solidity::frontend;
+
+void ImmutableValidator::analyze()
+{
+	m_readingOfImmutableAllowed = false;
+	m_initializationOfImmutableAllowed = true;
+
+	for (VariableDeclaration const* stateVar: m_currentContract.stateVariablesIncludingInherited())
+		if (stateVar->value())
+		{
+			stateVar->value()->accept(*this);
+			solAssert(m_initializedStateVariables.emplace(stateVar).second, "");
+		}
+
+	auto linearizeContracts = m_currentContract.annotation().linearizedBaseContracts  | boost::adaptors::reversed;
+
+	for (auto const* contract: linearizeContracts)
+		if (contract->constructor())
+			visitCallable(*contract->constructor());
+
+	m_initializationOfImmutableAllowed = false;
+
+	for (auto const* contract: linearizeContracts)
+		for (std::shared_ptr<InheritanceSpecifier> const inheritSpec: contract->baseContracts())
+			if (auto args = inheritSpec->arguments())
+				ASTNode::listAccept(*args, *this);
+
+	m_readingOfImmutableAllowed = true;
+
+	for (auto const* contract: linearizeContracts)
+	{
+		for (auto funcDef: contract->definedFunctions())
+			visitCallable(*funcDef);
+
+		for (auto modDef: contract->functionModifiers())
+			visitCallable(*modDef);
+	}
+
+	checkAllVariablesInitialized(m_currentContract.location());
+}
+
+bool ImmutableValidator::visit(FunctionDefinition const& _functionDefinition)
+{
+	return analyseCallable(_functionDefinition);
+}
+
+bool ImmutableValidator::visit(ModifierDefinition const& _modifierDefinition)
+{
+	return analyseCallable(_modifierDefinition);
+}
+
+bool ImmutableValidator::visit(MemberAccess const& _memberAccess)
```

Comment:
I'm wondering if it would be better to not implement any visitor for the expression nodes, but instead override the `visitNode(ASTNode const& _node)` function and check if `_node` is an expression.

Reply:
(although implementing the visitor for Identifier makes sense)

Reply:
Ah sorry, `referencedDeclaration` only exists for `MemberAccess` and `Identifier`, so it's fine as it is!